crates/core/src/event/extend.rs

```diff
@@ -0,0 +1,242 @@
+        CompositeEvent::new(self, data)
+    }
+}
+
+/// Event composed of various other event extensions.
+#[derive(Clone, Debug)]
+pub struct CompositeEvent<DATA, E> {
+    base_event: E,
+    data: DATA,
+}
+
+impl<E, DATA> CompositeEvent<DATA, E> {
+    /// Create a new composed event.
+    pub const fn new(base_event: E, data: DATA) -> Self {
+        Self { base_event, data }
+    }
+}
+
+impl<E, DATA> From<CompositeEvent<DATA, E>> for Event
+where
+    E: Into<Event>,
+    DATA: ExtendEvent,
+{
+    #[inline]
+    fn from(composite: CompositeEvent<DATA, E>) -> Event {
+        let CompositeEvent { base_event, data } = composite;
+
+        let mut base_event = base_event.into();
+        data.extend_event(&mut base_event);
+
+        base_event
+    }
+}
+
+/// Extend an [event](Event) with additional fields.
+pub trait ExtendEvent {
+    /// Add additional fields to the specified `event`.
+    fn extend_event(self, event: &mut Event);
+}
+
+/// Leaves an [`Event`] as is.
+pub struct WithNoOp;
+
+impl ExtendEvent for WithNoOp {
+    #[inline]
+    fn extend_event(self, _: &mut Event) {}
+}
+
+/// Extend an [`Event`] with block height information.
+pub struct WithBlockHeight(pub BlockHeight);
+
+impl ExtendEvent for WithBlockHeight {
+    #[inline]
+    fn extend_event(self, event: &mut Event) {
+        let Self(height) = self;
+        event["height"] = height.to_string();
+    }
+}
+
+/// Extend an [`Event`] with transaction hash information.
+pub struct WithTxHash(pub Hash);
+
+impl ExtendEvent for WithTxHash {
+    #[inline]
+    fn extend_event(self, event: &mut Event) {
+        let Self(hash) = self;
+        event["hash"] = hash.to_string();
+    }
+}
+
+/// Extend an [`Event`] with log data.
+pub struct WithLog(pub String);
+
+impl ExtendEvent for WithLog {
+    #[inline]
+    fn extend_event(self, event: &mut Event) {
+        let Self(log) = self;
+        event["log"] = log;
+    }
+}
+
+/// Extend an [`Event`] with info data.
+pub struct WithInfo(pub String);
+
+impl ExtendEvent for WithInfo {
+    #[inline]
+    fn extend_event(self, event: &mut Event) {
+        let Self(info) = self;
+        event["info"] = info;
+    }
+}
+
+/// Extend an [`Event`] with `is_valid_masp_tx` data.
+pub struct WithValidMaspTx(pub usize);
+
+impl ExtendEvent for WithValidMaspTx {
+    #[inline]
+    fn extend_event(self, event: &mut Event) {
+        let Self(masp_tx_index) = self;
+        event["is_valid_masp_tx"] = masp_tx_index.to_string();
+    }
+}
+
+/// Defer the extension of an [`Event`] to other
+/// implementations of [`ExtendEvent`].
+pub struct WithDeferredData<THIS, NEXT> {
+    this: THIS,
+    next: NEXT,
+}
+
+impl<THIS> WithDeferredData<THIS, WithNoOp> {
+    /// Begin composing a batch of [`ExtendEvent`] implementations.
+    pub const fn begin(data: THIS) -> Self {
+        Self {
+            this: data,
+            next: WithNoOp,
+        }
+    }
+}
+
+impl<THIS, NEXT> WithDeferredData<THIS, NEXT> {
```

Comment:
I really don't understand the need for this and I find it hard to read. It seems this could also be handled with extends.

Reply:
same reasoning here, albeit I admit the usefulness of `WithDeferredData` might be way more questionable lmao